File: 395_Longest_Substring_with_At_Least_K_Repeating_Characters/longestSubstring.py
class Solution(object):
    def longestSubstring(self, s, k):
        count = {c:0 for c in s}
        n = len(s)
        for i in xrange(n):
            if (1 + count[s[i]]) < k:
                if (n - i - 1) < k:
                    break
                else:
                    count[s[i]] += 1
            else:
                count[s[i]] += 1

File: 395_Longest_Substring_with_At_Least_K_Repeating_Characters/longestSubstring.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

class Solution {
    public:
        int longestSubstring(string s, int k) {
            int N = s.size();
            int maxL = 0;
            if(k > N) return 0;
            for(int len = N; len >= k; len--) {
                for(int left = 0; left < (N - len + 1); left++) {
                    int right = left + len;
                    map<char, int> count;
                    for(int m = left; m < right; m++) {
                        count[s[m]]++;
                    }
                    map<char, int>::iterator it;
                    for(it = count.begin(); it != count.end(); it++)
                        if(it -> second < k) break;
                    if(it == count.end()) return len;
                }
            }

            return 0;
        }
};

int main() {
    string s = "ababbc";
    int k = 2;
    Solution sol;
    cout << sol.longestSubstring(s, k) << endl;
    return 0;
}
